### PR TITLE
Add ability to specify used properties to avoid NLS search marking them

### PR DESCRIPTION
**Pull request by jjohnstn**
- add enhancement to be able to specify XXXX.usedproperties file which contains property keys known to be used (e.g. a key that gets dynamically created where it is used such as in ResourceAction
- modify NLSSearchResultRequestor to read .usedproperties files if they exist and keep a set of used properties
- modify NLSSearchQuery to check if an unused key has actually been specified as used via the .usedproperties file
- add new test to NLSSearchTest
- for #555

## What it does

Adds the ability to specify that NLS property keys are known to be used and should be ignored by the "Find broken NLS strings" action.  It does this by allowing properties from XXXX.properties to be added to XXXX.usedproperties.  The NLS search will mark such properties as used and will not report them.

## How to test

A new test has been added to NLSSearchTest which shows how it works.

## Author checklist

- [X] I have thoroughly tested my changes
- [X] The change is following the [coding conventions](https://wiki.eclipse.org/Platform/How_to_Contribute#Coding_Conventions)
- [X] I have signed the [Eclipse Contributor Agreement (ECA)](https://www.eclipse.org/legal/ECA.php)

### CODE REVIEW

**Comment by jjohnstn** [comment]
@iloveeclipse Any issues with this enhancement?  There are property keys used in JDT which are dynamically put together (e.g. by a ResourceAction() class which adds _label, _page, _xxxx etc.. to a prefix).  Rather than commenting them, I have added logic which will look for a .usedproperties file which marks any properties there as known to be used and ignored by the search.  Admittedly, comments could be placed in the .properties file, but it would be better to just not show any errors.  I made it work as a properties file so a user can simply copy the original file and delete those properties that are not protected.

**Comment by jjohnstn** [comment]
I will post a patch for 4.29 M1 which will remove the extraneous JDT UI messages for #555.  I have created the .usedproperties files already and can post them for 4.28 M3.

**Comment by iloveeclipse** [comment]
> Any issues with this enhancement

Jeff, the description makes sense and code wise seem to be OK. However, from the domain point of view I'm not experienced with NLS code at all, so can't say if there is a better or other way to achieve what you wanted to achieve here.

File: org.eclipse.jdt.ui/ui refactoring/org/eclipse/jdt/internal/ui/refactoring/nls/search/NLSSearchResultRequestor.java [hunk 1]
```diff
@@ -401,25 +399,30 @@ private int findPropertyNameStartPosition(String propertyName) {
...
-			} catch (IOException ex) {
-			}
 			reportDuplicateKeys(duplicateKeys);
 		}
+		if (!"properties".equalsIgnoreCase(fPropertiesFile.getFileExtension())) { //$NON-NLS-1$
+			return;
+		}
+		String propertyFileName= fPropertiesFile.getName();
+		String ignorePropertyFileName=
+				propertyFileName.substring(0, propertyFileName.length() - ".properties".length()).concat(".usedproperties"); //$NON-NLS-1$ //$NON-NLS-2$
```
[the remark this answers is not in the document]
Comment (jjohnstn) [reply org.eclipse.jdt.ui/ui refactoring/org/eclipse/jdt/internal/ui/refactoring/nls/search/NLSSearchResultRequestor.java:417]: Makes sense.

**Comment by jjohnstn** [comment]
@iloveeclipse Ok, thanks.  I just wanted to make sure there wasn't something I missed.  In most cases, it will never be used but it makes continuous checking of our projects feasible once messages are properly marked.

**Pull request merged**